bluefog/torch/optimizers.py

```diff
@@ -462,6 +463,343 @@ def step(self, closure=None):
+            state['square_avg'] = torch.zeros_like(p.data, memory_format=torch.preserve_format)
+            if param_group['momentum'] > 0:
+                state['momentum_buffer'] = torch.zeros_like(
+                    p.data, memory_format=torch.preserve_format)
+            if param_group['centered']:
+                state['grad_avg'] = torch.zeros_like(p.data, memory_format=torch.preserve_format)
+
+        square_avg = state['square_avg']
+        alpha = param_group['alpha']
+
+        state['step'] += 1
+        if param_group['weight_decay'] != 0:
+            grad = grad.add(p.data, alpha=param_group['weight_decay'])
+
+        square_avg.mul_(alpha).addcmul_(grad, grad, value=1 - alpha)
+
+        if param_group['centered']:
+            grad_avg = state['grad_avg']
+            grad_avg.mul_(alpha).add_(grad, alpha=1 - alpha)
+            avg = square_avg.addcmul(grad_avg, grad_avg, value=-1).sqrt_().add_(param_group['eps'])
+        else:
+            avg = square_avg.sqrt().add_(param_group['eps'])
+
+        if param_group['momentum'] > 0:
+            buf = state['momentum_buffer']
+            buf.mul_(param_group['momentum']).addcdiv_(grad, avg)
+            p.data.add_(buf, alpha=-param_group['lr'])
+        else:
+            p.data.addcdiv_(grad, avg, alpha=-param_group['lr'])
+
+    def _adagrad_step(self, p, grad, param_group):
+        state = self.state[p]
+        state['step'] += 1
+
+        if param_group['weight_decay'] != 0:
+            if grad.is_sparse:
+                raise RuntimeError("weight_decay option is not compatible with sparse gradients")
+            grad = grad.add(p.data, alpha=param_group['weight_decay'])
+
+        clr = param_group['lr'] / (1 + (state['step'] - 1) * param_group['lr_decay'])
+
+        if grad.is_sparse:
+            grad = grad.coalesce()  # the update is non-linear so indices must be unique
+            grad_indices = grad._indices()
+            grad_values = grad._values()
+            size = grad.size()
+
+            def make_sparse(values):
+                constructor = grad.new
+                if grad_indices.dim() == 0 or values.dim() == 0:
+                    return constructor().resize_as_(grad)
+                return constructor(grad_indices, values, size)
+            state['sum'].add_(make_sparse(grad_values.pow(2)))
+            std = state['sum'].sparse_mask(grad)
+            std_values = std._values().sqrt_().add_(param_group['eps'])
+            p.data.add_(make_sparse(grad_values / std_values), alpha=-clr)
+        else:
+            state['sum'].addcmul_(1, grad, grad)
+            std = state['sum'].sqrt().add_(param_group['eps'])
+            p.data.addcdiv_(grad, std, value=-clr)
+
+    def _adadelta_step(self, p, grad, param_group):
+        if grad.is_sparse:
+            raise RuntimeError('Adadelta does not support sparse gradients')
+        state = self.state[p]
+
+        # State initialization
+        if len(state) == 0:
```

Comment:
Seems that the style of these overridden step function is somewhat different. Like this one is 'len(state)==0' (the one in the PyTorch source), but previous use 'not state'; or the order in add_, addcmul_, etc.

Reply:
I should use len(state)==0 though the pylint doesn't recommend this.